python/tests/test_adjoint_solver.py

```diff
@@ -943,6 +943,123 @@ def test_mode_source_bandwidth(self):
                 self.assertLessEqual(rel_err, tol)
                 print(f"PASSED: fwidth={fwidth:.5f}, m={m}, err={rel_err:.10f}")
 
+    def test_periodic_design(self):
+        """Verifies that lengthscale constaint functions are invariant when
+        a design pattern is shifted along periodic directions."""
+        print("*** TESTING PERIODIC DESIGN ***")
+
+        # shifted design patterns
+        idx_row_shift, idx_col_shift = int(0.19 * self.Nx), int(0.28 * self.Ny)
+        p = self.p.reshape(self.Nx, self.Ny)
+        p_row_shift = np.vstack((p[idx_row_shift:, :], p[0:idx_row_shift, :]))
+        p_col_shift = np.hstack((p[:, idx_col_shift:], p[:, 0:idx_col_shift]))
+
+        eta_e = 0.55
+        eta_d = 1 - eta_e
+        beta, eta = 10, 0.5
+        radius, c = 0.3, 400
+        places = 18
```

Comment:
This seems way too low (smaller than machine precision). Are you sure this test is actually doing anything with the tolerance set so low?

Reply:
By making the tolerance so low it's basically checking that the value is invariant to machine precision.

Reply:
But I agree that the `places` should probably be smaller.

Reply:
@mawc2019 says that the absolute magnitude of the constraint functions is about `1e-5`, so this is asking it to match to 13 digits.   So this is fine.